src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -962,17 +962,23 @@ async Task<IEnumerable<IAccount>> ITokenCacheInternal.GetAccountsAsync(Authentic
             // Add WAM accounts stored in MSAL's cache - for which we do not have an RT
             if (requestParameters.AppConfig.IsBrokerEnabled && ServiceBundle.PlatformProxy.BrokerSupportsWamAccounts)
             {
-                foreach (MsalAccountCacheItem wamAccountCache in accountCacheItems.Where(
-                    acc => acc.WamAccountIds != null &&
-                    acc.WamAccountIds.ContainsKey(requestParameters.AppConfig.ClientId)))
+                foreach (MsalAccountCacheItem cachedAccount in accountCacheItems)
                 {
-                    var wamAccount = new Account(
-                        wamAccountCache.HomeAccountId,
-                        wamAccountCache.PreferredUsername,
-                        environment,
-                        wamAccountCache.WamAccountIds);
+                    if (!clientInfoToAccountMap.ContainsKey(cachedAccount.HomeAccountId) &&
+                        cachedAccount.WamAccountIds != null &&
```

Comment:
Thought about adding a home tenant filter here to return only home accounts with tenant profiles, as per guest account spec. However, that breaks a scenario when (1) there's only one guest account in the cache retrieved from WAM and (2) List OS Accounts filter is false. In that scenario the filter would filter that account out and return nothing. We might need to refactor some WAM code related to this because per spec when querying broker for a guest account, it will actually return a home account.

Reply:
Hmm, we have an algorithm to resolve guest account from WAM, i.e. we can always find the client_info. It's pretty complex, i.e. sometimes we'll have to actually make a token request.

Reply:
So for like an ATI with a guest authority, we do use client info and home account ID as a partition key but the account object represents guest account.